src/Microsoft.Data.Analysis/DataFrameWindow.Rolling.PrimitiveDataFrameColumn.cs

```diff
@@ -0,0 +1,213 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.Data.Analysis
+{
+    /// <summary>
+    /// A window for <see cref="PrimitiveDataFrameColumn{T}"/> to support rolling window operations
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class PrimitiveDataFrameColumnRollingWindow<T> : DataFrameColumnWindow
+        where T : unmanaged
+    {
+        private int _windowSize;
+        private PrimitiveDataFrameColumn<T> _currentColumn;
+        private long _currentIndex;
+
+        internal PrimitiveDataFrameColumnRollingWindow(int windowSize, PrimitiveDataFrameColumn<T> currentColumn)
+        {
+            _windowSize = windowSize;
+            _currentColumn = currentColumn;
+        }
+
+        /// <summary>
+        /// Apply <paramref name="func"/> over a rolling window of values. 
+        /// </summary>
+        /// <typeparam name="U">The returned result for each rolling window</typeparam>
+        /// <param name="func">The delegate to apply for each window. The first parameter is a linked list of values in the current window. The second parameter is the current index in the column. The current window ends at the current index and include values preceeding it</param>
+        /// <returns>The result computed for each rolling window</returns>
+        public PrimitiveDataFrameColumn<U> Apply<U>(Func<LinkedList<T?>, long, U?> func)
+            where U : unmanaged
+        {
+            return _currentColumn.ApplyRollingFunc(func, _windowSize);
+        }
+
+        /// <summary>
+        /// Finds the min of the values in this window
+        /// </summary>
+        /// <returns>Returns a DataFrame containing the min in each window</returns>
+        public new PrimitiveDataFrameColumn<T> Min()
+        {
+            _currentIndex = 0;
+
+            LinkedList<KeyValuePair<double, long>> sortedList = new LinkedList<KeyValuePair<double, long>>();
```

Comment:
Can you init these locals insided the lambda to avoid the closure (allocation)?

Reply:
But these are not local for each iteration in `Apply` though? `sortedList` needs to be global for the algorithm to work. If I moved `sortedList` inside the lambda, it would be allocated for each iteration in `Apply`?

Reply:
Ah, I didn't realize that `_currentColumn.Apply<T>` is an iteration. The you are right.